device-connectors/src/testflinger_device_connectors/devices/zapper/__init__.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (C) 2024 Canonical
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""
+Package containing modules for implementing Zapper-driven device connectors.
+
+Modules inheriting from the provided abstract class will run Zapper-driven
+provisioning procedures via Zapper API. The provisioning logic is implemented
+in the Zapper codebase and the connector serves as a pre-processing step,
+validating the configuration and preparing the API arguments.
+"""
+
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Tuple
+
+import rpyc
+import yaml
+
+import testflinger_device_connectors
+from testflinger_device_connectors import logmsg
+from testflinger_device_connectors.devices import (
+    DefaultDevice,
+    RecoveryError,
+    catch,
+)
+
+
+class ZapperConnector(ABC, DefaultDevice):
+    """
+    Abstract base class defining a common interface for Zapper-driven
+    device connectors.
+    """
+
+    PROVISION_METHOD = ""  # to be defined in the implementation
+    ZAPPER_REQUEST_TIMEOUT = 60 * 90
+    ZAPPER_SERVICE_PORT = 60000
+
+    @catch(RecoveryError, 46)
+    def provision(self, args):
+        """Method called when the command is invoked."""
+        with open(args.config) as configfile:
+            config = yaml.safe_load(configfile)
+        testflinger_device_connectors.configure_logging(config)
+
+        (api_args, api_kwargs) = self._validate_configuration(
+            args.config, args.job_data
+        )
+
+        logmsg(logging.INFO, "BEGIN provision")
+        logmsg(logging.INFO, "Provisioning device")
+
+        self._run(args.config["controller_host"], *api_args, **api_kwargs)
+
+        logmsg(logging.INFO, "END provision")
```

Comment:
This behaviour has been changed to logger.info in #197

Reply:
ah, time to rebase